Checks against projected future values

bittrance opened the pull request:
Support alerting against a linear projection into the future of the time series retrieved from Graphite. E.g. `--projection 30min` means that a future value will be calculated using linear regression and that value will be offered to nagios_check for evaluation against warning/critical thresholds. The check also presents the p-value of that projection so that ops people has a chance of judging whether the alarm is the result of noisy data.

Using a simple projection is superior in some use cases where the rate of change is relatively stable, because it saves us from having to estimate the rate of change. Instead we can just say e.g.: "Give me 6 hours advance notice on the disk filling up."

Commit:
Allow controlling how retrieved datapoints are processed

Commit:
Check warn/crit against a linear projection of retrieved data series into the future

iconara commented on check_graphite.gemspec:
```diff
@@ -20,6 +20,7 @@ Gem::Specification.new do |s|
 
   # specify any dependencies here; for example:
   # s.add_development_dependency "rspec"
+  s.add_runtime_dependency "linear-regression"
```
How much code would it add if the linear regression code was inlined (or rewritten in this gem)? I'm asking mostly so that we don't add dependencies unless absolutely necessary.

bittrance replied:
Hm, philosophy. Define absolutely necessary. (If it added significant code, I would not have used the external module.) You are right however that the module needs to have its version fixed in the gemspec.

iconara commented on lib/attime.rb:
```diff
@@ -0,0 +1,61 @@
+    raise "Unparseable time period #{text}" unless match
+    sign, scalar, unit = match[1..3]
+    raise "Missing scalar in time #{text}" if scalar == ""
+    raise "Missing unit in #{text}" if unit == ""
```
Use `scalar.empty?` instead of comparing to an empty string.

bittrance replied:
Fixed.

iconara commented on lib/attime.rb:
```diff
@@ -0,0 +1,61 @@
+      lookup(x)
+    else
+      x
+    end
```
I would rewrite like this:

```ruby
case x
when String
  lookup(x)
when nil
  raise "Bad unit #{unit}"
else
  x
end
```

to avoid a very easy-to-overlook trailing `unless` on the line just before the `if`.

bittrance replied:
Absolutely. Thx.

iconara commented on lib/attime.rb:
```diff
@@ -0,0 +1,61 @@
+    "year" => 365 * DAY,
+    "years" => "year",
+  }
+  EXPR = /([+-])?([0-9]*)(.*)/
```
Shouldn't this be `/\A([+-])?([0-9]*)(.*)\Z/` (i.e. anchored at the start and end of the string so that "1s 2d 3y" doesn't match?

bittrance replied:
The current implementation will yield "Bad unit s 2d 3y" which I feel is a good enough error message for this. `.*` at end is already anchored to end of input (unless there are newlines). However, the current implementation does not handle floats: `1.5days` => `Bad unit .5days` which are easier to support than fail intelligently. Fixed.

iconara commented on lib/attime.rb:
```diff
@@ -0,0 +1,61 @@
+    raise "Missing scalar in time #{text}" if scalar == ""
+    raise "Missing unit in #{text}" if unit == ""
+    time + Integer(scalar) * lookup(unit) * (sign == "-" ? -1 : 1)
+  end
```
I'm on the fence about trailing `if`s and `unless`s. I think they have their place and can add readability, but using them too much in a method, especially with calculations in between, IMO makes the code noisy.

Why does the expression even allow an empty scalar and/or unit? Does it really add that much to the error message to differentiate between the cases? Could it be enough to just say that the input is malformed when the regex doesn't match?

bittrance replied:
Yes, giving intelligent error messages on option parsing for data where it is not trivial to know what is possible is critical.

iconara commented on lib/check_graphite/version.rb:
```diff
@@ -1,3 +1,3 @@
 module CheckGraphite
-  VERSION = "0.2.5"
+  VERSION = "0.3.0.pre1"
```
Version bumping is better to do on master after a merge. You never know how long it takes before something gets merge, maybe master has raced past the version in the branch before it's merged.

bittrance replied:
While I agree on the principle, I hold that it is an absolute rule that you must never publish software without documenting the fact and version that you have done so. Since this branch is ultimately destined for upstream repo, I did not want to merge it to burcorp master, and so the version is bumped here and tagged accordingly. This is acceptable for release candidates but not for finals.

iconara commented on spec/attime_spec.rb:
```diff
@@ -0,0 +1,31 @@
+  it("2months") { should be_within(0.9).of(timedelta(60 * 86400)) }
+  it("12months") { should be_within(0.9).of(timedelta(360 * 86400)) }
+  it("1year") { should be_within(0.9).of(timedelta(365 * 86400)) }
+  it("2years") { should be_within(0.9).of(timedelta(730 * 86400)) }
```
Clever, but if you want it short, and IMO more readable:

```ruby
examples = {
  "10days" => 10 * 86400,
  "0days" => 0,
  …
  "2years" => 730 * 86400,
}

examples.each do |expression, expected_result|
  it "parses #{expression}" do
    Graphite.attime(expression).should eq(expected_result)
  end
end
```

Or with a modern RSpec (not sure if that's possible):

```
it 'parses a lot of different expressions', :aggregate_failures do
  examples.each do |expression, expected_result|
    expect(described_class.attime(expression)).to eq(expected_result)
  end
end
```

bittrance replied:
You would actually get `"10days" => be_within(0.9).of(10 * 86400)`. The proposed style breaks the rules that tests should be reasonably independent and I do not think it really improves readability.

bittrance replied:
s/independent/self-sufficient/

iconara commented on spec/projection_spec.rb:
```diff
@@ -0,0 +1,80 @@
+require 'check_graphite/projection'
+require 'nagios_check'
+
+describe describe CheckGraphite::Projection do
```
Double `describe`

bittrance replied:
Strange that you can do that. Fixed.

iconara commented on spec/projection_spec.rb:
```diff
@@ -0,0 +1,80 @@
+    end.new
+  end
+
+  describe 'primary value' do
```
`describe` should be used for classes and methods and `context` for sections within `describe`s that require extra qualification. The description of `context` should generally start with "with", "when" (or "and" if they are nested within another `context`) so that the full description of a test (the concatenation of the strings of all `describe`, `context` and `it` descriptions) becomes a readable sentence (or as close as possible).

The descriptions of tests in this context will have the prefix "CheckGraphite::Projection primary value", which doesn't read very well. I guess you're thinking of "primary value" kind of like a method (if it were a method using `describe '#primary_value'` would be exactly how you'd write it), but it looks a bit weird when it's not actually a method.

Using `context 'when looking at the primary value'` would work better, for example.

The contexts below are well written, though.

bittrance replied:
I read "describe should be used for classes and methods and context for sections within describes that require extra qualification" as a mockist argument. To me `describe` is about a unit of functionality and a particular method can be split into any number of units. The current labels are atrocious tho. Fixed.

iconara commented on spec/projection_spec.rb:
```diff
@@ -0,0 +1,80 @@
+      let(:projection) { '2sec' }
+      let(:datapoints) { [[10,0], [9,1], [8,2]] }
+
+      it { should be_within(0.01).of(6) }
```
I'm not convinced by these description-less tests, I don't like the test descriptions. This one would become "should be within 0.01 of 6" for example. That doesn't tell me anything. It doesn't say why that is good or why it's right. Something like 'predicts the next value using a linear model' would better explain what's going on.

bittrance replied:
Indeed. Fixed.

iconara commented on spec/projection_spec.rb:
```diff
@@ -0,0 +1,80 @@
+    context 'given a constant series (where Pearson is undefined)' do
+      let(:projection) { '2sec' }
+      let(:datapoints) { [[10,0], [10,1], [10,2]] }
+      it { should eq('undefined') }
```
This is an example where the description-less syntax kind of works: 'CheckGraphite::Projection p-value given a constant series (where Pearson is undefined) should eq "undefined"'. However, adding a hand-crafted description would be even better: "… given a constant series (where Pearson is undefined) returns undefined", but it's marginal.

iconara commented on spec/projection_spec.rb:
```diff
@@ -0,0 +1,80 @@
+      )
+    end
+
+    it 'should output value, projection interval and p-value' do
```
I prefer not to start test descriptions with "should" (another reason I avoid RSpecs description-less tests). Tests show how things are, not how they should be. "should" implies uncertainty.

bittrance replied:
Indeed. Fixed.

iconara commented on spec/projection_spec.rb:
```diff
@@ -0,0 +1,80 @@
+    end
+  end
+
+  describe 'integration test' do
```
If these are integration tests I think they belong in another file, but since they stub the other side I don't really think they're integration tests and the description could be something that better explain what they test (rather than what they are), for example "when making a HTTP request to Graphite" or something like that.

bittrance replied:
The main problem here is that the test resides inside CheckGraphite::Projection at all. Rather, it is a test to make sure that projections are invoked. Even asserting on the output is a bit overreach. Changing this to a separate describe saying `when invoking graphite with --projection`.

iconara commented on lib/check_graphite/projection.rb:
```diff
@@ -0,0 +1,39 @@
+      p = Regression::CorrelationCoefficient.new(xs, ys).pearson
+      store_value options.name, value
+      # Unfortunately, nagios_check converts both primary and
+      # secondary values to float. Hence manual assignment instead:
```
I think this explanation belongs in a test and/or a commit message. Comments usually lie because they remain after the code is changed.

bittrance replied:
I disagree. While comments should be used sparingly, If I break or circumvent a contract, I need to explain why. If you try to change this to `store_value 'p-value', p.abs` you will get an exception from the `undefined` test. Creating a test with the title "make sure the contract is circumvented" is just guessing on the reason for a future breakage; there is no way a test can be that surgical.

If a developer disregards an immediately adjacent comment, that is no different from disregarding an immediately adjacent line of code. I don't really see how a review can objectively object to a useful comment when the review could equally have objected to that comment if it was orphaned.

iconara commented on lib/check_graphite/projection.rb:
```diff
@@ -0,0 +1,39 @@
+    def self.included(base)
+      base.on '--projection FUTURE_TIMEFRAME', :default => '2days' do |timeframe|
+        options.send('processor=', method(:projected_value))
+        options.send('timeframe=', timeframe)
```
The Ruby-ish thing to do is to use `:processor=` and `:timeframe=` (i.e. symbols not strings when referring to methods).

bittrance replied:
Indeed. Fixed.

iconara commented on spec/attime_spec.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  def timedelta(n)
+    Time.now + n
```
Let's hope there's no GC pauses during the test run... since `.attime` takes the current time as second argument I think you should have a `let :now` and use that here and in the `.attime` call to make sure that time doesn't elapse during the test run.

bittrance replied:
It hardly takes 0.9 secs to GC a test execution? But you are right, it makes for better specs. In fact, attime() should probably require the time since that is how it is used in production code.

iconara commented on spec/attime_spec.rb:
```diff
@@ -0,0 +1,31 @@
+require 'attime'
+
+describe '#attime' do
```
I think this should be:

```ruby
describe CheckGraphite do
  describe '.attime' do
```

or 

```ruby
describe 'CheckGraphite.attime' do
```

I.e. the description should include the module/class name, and in Ruby you refer to class/module methods as `.foo` and only instance methods as `#foo`.

bittrance replied:
Survived from a previous incarnation; changed to second version.

iconara commented on spec/projection_spec.rb:
```diff
@@ -0,0 +1,80 @@
+      })
+      c = CheckGraphite::Command.new
+      STDOUT.should_receive(:puts).with(match(/ze-name.*in 5min.*p-value=0.9/))
+      lambda { c.run }.should raise_error SystemExit
```
I kind of get why this is needed, but boy is this a brittle test... both the stub of stdout and then rescuing `SystemExit`... It's not very much less ugly, but perhaps `Kernel.should_receive(:exit).with(0)`?

bittrance replied:
This is copied from check_graphite_spec.rb. I do not approve of the style, but solving the i/o problem would require to introduce a way to inject a StringIO. `raise_error SystemExit` is not so fragile because you will probably notice if it is not satisfied (but `not_to raise_error SystemExit` would be essentially worthless).

iconara commented:
In general this is great 👍 , but I had a few comments.

Commit:
- lock linear-regression to 0.0.2
- support float timeframes e.g. --projection 1.5days
- style improvements as suggested by iconara on PR #1

bittrance commented:
Thanks for taking the time to review this. It was concieved in relative haste and is somewhat shoddy.

iconara commented on lib/check_graphite/projection.rb:
```diff
@@ -0,0 +1,39 @@
+       if v.nan?
+         'undefined'
+       else
+         BigDecimal.new(v, 3).to_s('F')
```
This thing needs a test. It needs to be documented why the formatting uses three significant digits (and not five, three or 99). Also, why not for example `sprintf('%.3f', v)` (because I assume this method only ever expects numbers in the 0.0-1.0 range)?

This line has actually been bugging me for a few days, but it was not until today when it just popped into my head that I realised why. It's one of those things that you kind of gloss over, like "yeah, sure and then there's some formatting". But it's not just formatting, it's very deliberate formatting, but with no test to motivate it.

bittrance replied:
sprintf() can only do number of decimals, but here we deal with numbers that can be any float, so there is no way to know how many decimals are relevant, if any. Thus we instead do number of significant digits. Why 3? It works for our use case; it will need to be configurable eventually, but I think it would be a rare use case that cares about changes < 1/1000. I don't think the helper merits being named a test unit in its own right. However, this means that the "integration" tests that tests against the full output should check that the number is formatted correctly. Fixed.

bittrance replied:
It turned out to make for messy rebases, so this is actually fixed on b2b1738 not in this PR.

iconara reviewed (changes requested):
I'm missing a test.
